# Exploring Hackers News Posts¶
In this project, I will compare two different types of posts from [Hacker News](https://news.ycombinator.com/), a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts i will explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

I will specifically compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?
It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions


# Introduction¶
First, we'll read in the data and remove the headers.




In [2]:
from csv import reader
opened_file=open("hacker_news.csv")
hn=reader(opened_file)
hn=list(hn)
print(hn[0:4])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']]


# Removing Headers from a List of Lists¶


In [3]:
headers=hn[0]
print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [4]:
hn=hn[1:]

In [5]:
print(hn[0:4])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


# Extracting Ask HN and Show HN Posts¶
First, i will identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.




In [32]:
# Identify posts that begin with either `Ask HN` or `Show HN` and separate the data into different lists.

ask_posts=[]
show_posts=[]
other_posts=[]
for row in hn:
    title=row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [7]:
print(ask_posts[0:2])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43']]


# Calculating the Average Number of Comments for Ask HN and Show HN Posts¶
Now that i separated ask posts and show posts into different lists, i'll calculate the average number of comments each type of post receives.




In [37]:
# Calculate the average number of comments `Ask HN` posts receive.

total_ask_comments=0
for row in ask_posts:
    comments=int(row[4])
    total_ask_comments+=comments
avg_ask_comments=total_ask_comments/len(ask_posts)
    

In [9]:
print(len(ask_posts))

1744


In [10]:
total_show_comments=0
for row in show_posts:
    comments=int(row[4])
    total_show_comments+=comments
avg_show_comments=total_show_comments/len(show_posts)
    

In [11]:
print(avg_show_comments)

10.31669535283993


In [12]:
print(avg_ask_comments)

14.038417431192661


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, I'll focus our remaining analysis just on these posts.



# Finding the Amount of Ask Posts and Comments by Hour Created¶
Next, I'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, I'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, I'll calculate the average amount of comments ask posts created at each hour of the day receive.



In [13]:

import datetime as dt
date_1_str = "December 24, 1984"
date_1_dt = dt.datetime.strptime(date_1_str, "%B %d, %Y")
print(date_1_dt)

1984-12-24 00:00:00


In [14]:
result_list=[]
for row in ask_posts:
    created_date=row[6]
    num_comments=int(row[4])
    result_list.append([created_date,num_comments])

In [15]:
print(result_list[0:3])

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1]]


In [38]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received.

import datetime as dt
counts_by_hour={}
comments_by_hour={}
date_format = "%m/%d/%Y %H:%M"
for row in result_list:
    date=row[0]
    hour=dt.datetime.strptime(date,date_format).strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour]=1
        comments_by_hour[hour]=row[1]
    else:
        counts_by_hour[hour]+=1
        comments_by_hour[hour]+=row[1]
        
    

In [23]:
print(comments_by_hour)

{'05': 464, '03': 421, '06': 397, '19': 1188, '07': 267, '08': 492, '23': 543, '01': 683, '21': 1745, '04': 337, '22': 479, '13': 1253, '17': 1146, '10': 793, '09': 251, '20': 1722, '00': 447, '16': 1814, '15': 4477, '18': 1439, '02': 1381, '11': 641, '14': 1416, '12': 687}


In [24]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

avg_by_hour


[['05', 10.08695652173913],
 ['03', 7.796296296296297],
 ['06', 9.022727272727273],
 ['19', 10.8],
 ['07', 7.852941176470588],
 ['08', 10.25],
 ['23', 7.985294117647059],
 ['01', 11.383333333333333],
 ['21', 16.009174311926607],
 ['04', 7.170212765957447],
 ['22', 6.746478873239437],
 ['13', 14.741176470588234],
 ['17', 11.46],
 ['10', 13.440677966101696],
 ['09', 5.5777777777777775],
 ['20', 21.525],
 ['00', 8.127272727272727],
 ['16', 16.796296296296298],
 ['15', 38.5948275862069],
 ['18', 13.20183486238532],
 ['02', 23.810344827586206],
 ['11', 11.051724137931034],
 ['14', 13.233644859813085],
 ['12', 9.41095890410959]]

In [25]:
swap_avg_by_hour=[]
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])

In [26]:
print(swap_avg_by_hour)

[[10.08695652173913, '05'], [7.796296296296297, '03'], [9.022727272727273, '06'], [10.8, '19'], [7.852941176470588, '07'], [10.25, '08'], [7.985294117647059, '23'], [11.383333333333333, '01'], [16.009174311926607, '21'], [7.170212765957447, '04'], [6.746478873239437, '22'], [14.741176470588234, '13'], [11.46, '17'], [13.440677966101696, '10'], [5.5777777777777775, '09'], [21.525, '20'], [8.127272727272727, '00'], [16.796296296296298, '16'], [38.5948275862069, '15'], [13.20183486238532, '18'], [23.810344827586206, '02'], [11.051724137931034, '11'], [13.233644859813085, '14'], [9.41095890410959, '12']]


Sorting and Printing Values from a List of Lists


In [28]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print(sorted_swap)

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


In [31]:

# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print("{}:{:.2f} average commenst per post".format(dt.datetime.strptime(hr,"%H").strftime("%H:%M"),avg))


Top 5 Hours for 'Ask HN' Comments
15:00:38.59 average commenst per post
02:00:23.81 average commenst per post
20:00:21.52 average commenst per post
16:00:16.80 average commenst per post
21:00:16.01 average commenst per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts/home), the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

# Conclusion
In this project, I analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, i recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.